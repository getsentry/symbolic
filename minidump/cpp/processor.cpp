#include "google_breakpad/processor/basic_source_line_resolver.h"
#include "google_breakpad/processor/minidump.h"
#include "google_breakpad/processor/minidump_processor.h"
#include "google_breakpad/processor/process_state.h"

#include "cpp/data_definitions.h"
#include "cpp/memstream.h"
#include "cpp/mmap_symbol_supplier.h"
#include "cpp/processor.h"

using google_breakpad::BasicSourceLineResolver;
using google_breakpad::Minidump;
using google_breakpad::MinidumpProcessor;
using google_breakpad::ProcessState;

process_state_t *process_minidump(const char *buffer,
                                  size_t buffer_size,
                                  symbol_entry_t *symbols,
                                  size_t symbol_count,
                                  int *result_out) {
    if (buffer == nullptr) {
        *result_out = google_breakpad::PROCESS_ERROR_MINIDUMP_NOT_FOUND;
        return nullptr;
    }

    ProcessState *state = new ProcessState();
    if (state == nullptr) {
        *result_out = -1;  // Memory allocation issue
        return nullptr;
    }

    BasicSourceLineResolver resolver;
    MmapSymbolSupplier supplier(symbol_count, symbols);
    MinidumpProcessor processor(&supplier, &resolver);

    imemstream in(buffer, buffer_size);
    Minidump minidump(in);
    if (!minidump.Read()) {
        *result_out = google_breakpad::PROCESS_ERROR_MINIDUMP_NOT_FOUND;
        return nullptr;
    }

    *result_out = processor.Process(&minidump, state);
    if (*result_out != google_breakpad::PROCESS_OK) {
        delete state;
        return nullptr;
    }

    return process_state_t::cast(state);
}
