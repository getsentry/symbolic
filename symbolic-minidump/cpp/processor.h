#ifndef SENTRY_PROCESSOR_H
#define SENTRY_PROCESSOR_H

#include <cstddef>
#include "cpp/data_structures.h"

#ifdef __cplusplus
extern "C" {
#endif

/// Data transfer object for symbols in memory
struct symbol_entry_t {
    /// The debug identifier of the code module these symbols are for
    const char *debug_identifier;

    /// Size of the buffer inside symbol_data
    const size_t symbol_size;

    /// Raw data of the symbol file passed to the symbolizer
    const char *symbol_data;
};

/// Reads a minidump from a memory buffer and processes it. Returns an owning
/// pointer to a process_state_t struct that contains loaded code modules and
/// call stacks of all threads of the process during the crash.
///
/// Processing the minidump can fail if the buffer is corrupted or does not
/// exit. The function will return NULL and an error code in result_out.
///
/// Release memory of the process state with process_state_delete.
process_state_t *process_minidump(const char *buffer,
                                  size_t buffer_size,
                                  symbol_entry_t *symbols,
                                  size_t symbol_count,
                                  int *result_out);

#ifdef __cplusplus
}
#endif

#endif
