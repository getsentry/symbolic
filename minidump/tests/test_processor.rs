use failure::Error;
use insta;

use symbolic_common::ByteView;
use symbolic_minidump::processor::ProcessState;

#[test]
fn process_minidump_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/linux/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_linux", &state);
    Ok(())
}

#[test]
fn process_minidump_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/macos/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_macos", &state);
    Ok(())
}

#[test]
fn process_minidump_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_windows", &state);
    Ok(())
}

#[test]
fn get_referenced_modules_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/linux/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("referenced_modules_linux", &state.referenced_modules());
    Ok(())
}

#[test]
fn get_referenced_modules_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/macos/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("referenced_modules_macos", &state.referenced_modules());
    Ok(())
}

#[test]
fn get_referenced_modules_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!(
        "referenced_modules_windows",
        &state.referenced_modules()
    );
    Ok(())
}
