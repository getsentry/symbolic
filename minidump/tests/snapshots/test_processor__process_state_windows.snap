---
created: "2019-02-20T15:42:12.012643Z"
creator: insta@0.6.3
source: minidump/tests/test_processor.rs
expression: "&state"
---
ProcessState {
    requesting_thread: 0,
    timestamp: 1521713273,
    crash_address: 69,
    crash_reason: "EXCEPTION_ACCESS_VIOLATION_WRITE",
    assertion: "",
    system_info: SystemInfo {
        os_name: "Windows NT",
        os_version: "10.0.14393",
        cpu_family: "x86",
        cpu_info: "GenuineIntel family 6 model 63 stepping 2",
        cpu_count: 2
    },
    threads: [
        CallStack {
            thread_id: 1636,
            frames: [
                StackFrame {
                    return_address: 2763325,
                    instruction: 2763325,
                    trust: Context,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 2752512,
                            size: 36864,
                            code_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe",
                            code_identifier: "5AB380779000",
                            debug_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb",
                            debug_identifier: "3249D99D0C4049318610F4E4FB0B69361"
                        }
                    )
                },
                StackFrame {
                    return_address: 2762960,
                    instruction: 2762959,
                    trust: FP,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 2752512,
                            size: 36864,
                            code_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe",
                            code_identifier: "5AB380779000",
                            debug_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb",
                            debug_identifier: "3249D99D0C4049318610F4E4FB0B69361"
                        }
                    )
                },
                StackFrame {
                    return_address: 1971644864,
                    instruction: 1971644863,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "ae131c67-27a7-4fa1-9916-b5a4aef41190",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1971388416,
                            size: 790528,
                            code_file: "C:\\Windows\\System32\\rpcrt4.dll",
                            code_identifier: "5A49BB75c1000",
                            debug_file: "wrpcrt4.pdb",
                            debug_identifier: "AE131C6727A74FA19916B5A4AEF411901"
                        }
                    )
                },
                StackFrame {
                    return_address: 1887764480,
                    instruction: 1887764479,
                    trust: Scan,
                    module: None
                },
                StackFrame {
                    return_address: 1891085888,
                    instruction: 1891085887,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "aec7ef2f-df4b-4642-a471-4c3e5fe8760a",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1891041280,
                            size: 151552,
                            code_file: "C:\\Windows\\System32\\dbgcore.dll",
                            code_identifier: "57898DAB25000",
                            debug_file: "dbgcore.pdb",
                            debug_identifier: "AEC7EF2FDF4B4642A4714C3E5FE8760A1"
                        }
                    )
                },
                StackFrame {
                    return_address: 1971388416,
                    instruction: 1971388415,
                    trust: Scan,
                    module: None
                },
                StackFrame {
                    return_address: 1971644864,
                    instruction: 1971644863,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "ae131c67-27a7-4fa1-9916-b5a4aef41190",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1971388416,
                            size: 790528,
                            code_file: "C:\\Windows\\System32\\rpcrt4.dll",
                            code_identifier: "5A49BB75c1000",
                            debug_file: "wrpcrt4.pdb",
                            debug_identifier: "AE131C6727A74FA19916B5A4AEF411901"
                        }
                    )
                },
                StackFrame {
                    return_address: 2765877,
                    instruction: 2765876,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 2752512,
                            size: 36864,
                            code_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe",
                            code_identifier: "5AB380779000",
                            debug_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb",
                            debug_identifier: "3249D99D0C4049318610F4E4FB0B69361"
                        }
                    )
                },
                StackFrame {
                    return_address: 2764183,
                    instruction: 2764182,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 2752512,
                            size: 36864,
                            code_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe",
                            code_identifier: "5AB380779000",
                            debug_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb",
                            debug_identifier: "3249D99D0C4049318610F4E4FB0B69361"
                        }
                    )
                },
                StackFrame {
                    return_address: 1963352772,
                    instruction: 1963352771,
                    trust: FP,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "d3474559-96f7-47d6-bf43-c176b2171e68",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1963261952,
                            size: 917504,
                            code_file: "C:\\Windows\\System32\\kernel32.dll",
                            code_identifier: "590285E9e0000",
                            debug_file: "wkernel32.pdb",
                            debug_identifier: "D347455996F747D6BF43C176B2171E681"
                        }
                    )
                },
                StackFrame {
                    return_address: 1998393209,
                    instruction: 1998393208,
                    trust: FP,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "971f98e5-ce60-41ff-b2d7-235bbeb34578",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1997996032,
                            size: 1585152,
                            code_file: "C:\\Windows\\System32\\ntdll.dll",
                            code_identifier: "59B0D8F3183000",
                            debug_file: "wntdll.pdb",
                            debug_identifier: "971F98E5CE6041FFB2D7235BBEB345781"
                        }
                    )
                },
                StackFrame {
                    return_address: 1998393156,
                    instruction: 1998393155,
                    trust: FP,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "971f98e5-ce60-41ff-b2d7-235bbeb34578",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1997996032,
                            size: 1585152,
                            code_file: "C:\\Windows\\System32\\ntdll.dll",
                            code_identifier: "59B0D8F3183000",
                            debug_file: "wntdll.pdb",
                            debug_identifier: "971F98E5CE6041FFB2D7235BBEB345781"
                        }
                    )
                }
            ]
        },
        CallStack {
            thread_id: 3580,
            frames: [
                StackFrame {
                    return_address: 1998455148,
                    instruction: 1998455148,
                    trust: Context,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "971f98e5-ce60-41ff-b2d7-235bbeb34578",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1997996032,
                            size: 1585152,
                            code_file: "C:\\Windows\\System32\\ntdll.dll",
                            code_identifier: "59B0D8F3183000",
                            debug_file: "wntdll.pdb",
                            debug_identifier: "971F98E5CE6041FFB2D7235BBEB345781"
                        }
                    )
                },
                StackFrame {
                    return_address: 1963352772,
                    instruction: 1963352771,
                    trust: FP,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "d3474559-96f7-47d6-bf43-c176b2171e68",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1963261952,
                            size: 917504,
                            code_file: "C:\\Windows\\System32\\kernel32.dll",
                            code_identifier: "590285E9e0000",
                            debug_file: "wkernel32.pdb",
                            debug_identifier: "D347455996F747D6BF43C176B2171E681"
                        }
                    )
                },
                StackFrame {
                    return_address: 1998393209,
                    instruction: 1998393208,
                    trust: FP,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "971f98e5-ce60-41ff-b2d7-235bbeb34578",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1997996032,
                            size: 1585152,
                            code_file: "C:\\Windows\\System32\\ntdll.dll",
                            code_identifier: "59B0D8F3183000",
                            debug_file: "wntdll.pdb",
                            debug_identifier: "971F98E5CE6041FFB2D7235BBEB345781"
                        }
                    )
                },
                StackFrame {
                    return_address: 1998393156,
                    instruction: 1998393155,
                    trust: FP,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "971f98e5-ce60-41ff-b2d7-235bbeb34578",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1997996032,
                            size: 1585152,
                            code_file: "C:\\Windows\\System32\\ntdll.dll",
                            code_identifier: "59B0D8F3183000",
                            debug_file: "wntdll.pdb",
                            debug_identifier: "971F98E5CE6041FFB2D7235BBEB345781"
                        }
                    )
                }
            ]
        },
        CallStack {
            thread_id: 2600,
            frames: [
                StackFrame {
                    return_address: 1998455148,
                    instruction: 1998455148,
                    trust: Context,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "971f98e5-ce60-41ff-b2d7-235bbeb34578",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1997996032,
                            size: 1585152,
                            code_file: "C:\\Windows\\System32\\ntdll.dll",
                            code_identifier: "59B0D8F3183000",
                            debug_file: "wntdll.pdb",
                            debug_identifier: "971F98E5CE6041FFB2D7235BBEB345781"
                        }
                    )
                },
                StackFrame {
                    return_address: 1963352772,
                    instruction: 1963352771,
                    trust: FP,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "d3474559-96f7-47d6-bf43-c176b2171e68",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1963261952,
                            size: 917504,
                            code_file: "C:\\Windows\\System32\\kernel32.dll",
                            code_identifier: "590285E9e0000",
                            debug_file: "wkernel32.pdb",
                            debug_identifier: "D347455996F747D6BF43C176B2171E681"
                        }
                    )
                },
                StackFrame {
                    return_address: 1998393209,
                    instruction: 1998393208,
                    trust: FP,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "971f98e5-ce60-41ff-b2d7-235bbeb34578",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1997996032,
                            size: 1585152,
                            code_file: "C:\\Windows\\System32\\ntdll.dll",
                            code_identifier: "59B0D8F3183000",
                            debug_file: "wntdll.pdb",
                            debug_identifier: "971F98E5CE6041FFB2D7235BBEB345781"
                        }
                    )
                },
                StackFrame {
                    return_address: 1998393156,
                    instruction: 1998393155,
                    trust: FP,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "971f98e5-ce60-41ff-b2d7-235bbeb34578",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1997996032,
                            size: 1585152,
                            code_file: "C:\\Windows\\System32\\ntdll.dll",
                            code_identifier: "59B0D8F3183000",
                            debug_file: "wntdll.pdb",
                            debug_identifier: "971F98E5CE6041FFB2D7235BBEB345781"
                        }
                    )
                }
            ]
        },
        CallStack {
            thread_id: 2920,
            frames: [
                StackFrame {
                    return_address: 1998451676,
                    instruction: 1998451676,
                    trust: Context,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "971f98e5-ce60-41ff-b2d7-235bbeb34578",
                                        appendix: 1
                                    }
                                }
                            ),
                            base_address: 1997996032,
                            size: 1585152,
                            code_file: "C:\\Windows\\System32\\ntdll.dll",
                            code_identifier: "59B0D8F3183000",
                            debug_file: "wntdll.pdb",
                            debug_identifier: "971F98E5CE6041FFB2D7235BBEB345781"
                        }
                    )
                }
            ]
        }
    ],
    modules: [
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
                        appendix: 1
                    }
                }
            ),
            base_address: 2752512,
            size: 36864,
            code_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.exe",
            code_identifier: "5AB380779000",
            debug_file: "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb",
            debug_identifier: "3249D99D0C4049318610F4E4FB0B69361"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "9c2a902b-6fdf-40ad-8308-588a41d572a0",
                        appendix: 1
                    }
                }
            ),
            base_address: 1887764480,
            size: 1331200,
            code_file: "C:\\Windows\\System32\\dbghelp.dll",
            code_identifier: "57898E12145000",
            debug_file: "dbghelp.pdb",
            debug_identifier: "9C2A902B6FDF40AD8308588A41D572A01"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "bf5257f7-8c26-43dd-9bb7-901625e1136a",
                        appendix: 1
                    }
                }
            ),
            base_address: 1889140736,
            size: 442368,
            code_file: "C:\\Windows\\System32\\msvcp140.dll",
            code_identifier: "589ABC846c000",
            debug_file: "msvcp140.i386.pdb",
            debug_identifier: "BF5257F78C2643DD9BB7901625E1136A1"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "8daf7773-372f-460a-af38-944e193f7e33",
                        appendix: 1
                    }
                }
            ),
            base_address: 1889599488,
            size: 598016,
            code_file: "C:\\Windows\\System32\\apphelp.dll",
            code_identifier: "57898EEB92000",
            debug_file: "apphelp.pdb",
            debug_identifier: "8DAF7773372F460AAF38944E193F7E331"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "aec7ef2f-df4b-4642-a471-4c3e5fe8760a",
                        appendix: 1
                    }
                }
            ),
            base_address: 1891041280,
            size: 151552,
            code_file: "C:\\Windows\\System32\\dbgcore.dll",
            code_identifier: "57898DAB25000",
            debug_file: "dbgcore.pdb",
            debug_identifier: "AEC7EF2FDF4B4642A4714C3E5FE8760A1"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "0ed80a50-ecda-472b-86a4-eb6c833f8e1b",
                        appendix: 1
                    }
                }
            ),
            base_address: 1892024320,
            size: 81920,
            code_file: "C:\\Windows\\System32\\VCRUNTIME140.dll",
            code_identifier: "589ABC7714000",
            debug_file: "vcruntime140.i386.pdb",
            debug_identifier: "0ED80A50ECDA472B86A4EB6C833F8E1B1"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "147c51fb-7ca1-408f-85b5-285f2ad6f9c5",
                        appendix: 1
                    }
                }
            ),
            base_address: 1941569536,
            size: 40960,
            code_file: "C:\\Windows\\System32\\CRYPTBASE.dll",
            code_identifier: "57899141a000",
            debug_file: "cryptbase.pdb",
            debug_identifier: "147C51FB7CA1408F85B5285F2AD6F9C51"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "51e432b1-0450-4b19-8ed1-6d4335f9f543",
                        appendix: 1
                    }
                }
            ),
            base_address: 1941635072,
            size: 126976,
            code_file: "C:\\Windows\\System32\\sspicli.dll",
            code_identifier: "59BF30E31f000",
            debug_file: "wsspicli.pdb",
            debug_identifier: "51E432B104504B198ED16D4335F9F5431"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "0c799483-b549-417d-8433-4331852031fe",
                        appendix: 1
                    }
                }
            ),
            base_address: 1942421504,
            size: 487424,
            code_file: "C:\\Windows\\System32\\advapi32.dll",
            code_identifier: "5A49BB7677000",
            debug_file: "advapi32.pdb",
            debug_identifier: "0C799483B549417D84334331852031FE1"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "6f6409b3-d520-43c7-9b2f-62e00bfe761c",
                        appendix: 1
                    }
                }
            ),
            base_address: 1942945792,
            size: 778240,
            code_file: "C:\\Windows\\System32\\msvcrt.dll",
            code_identifier: "57899155be000",
            debug_file: "msvcrt.pdb",
            debug_identifier: "6F6409B3D52043C79B2F62E00BFE761C1"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "6f6a05dd-0a80-478b-a419-9b88703bf75b",
                        appendix: 1
                    }
                }
            ),
            base_address: 1950679040,
            size: 266240,
            code_file: "C:\\Windows\\System32\\sechost.dll",
            code_identifier: "598942C741000",
            debug_file: "sechost.pdb",
            debug_identifier: "6F6A05DD0A80478BA4199B88703BF75B1"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "d3474559-96f7-47d6-bf43-c176b2171e68",
                        appendix: 1
                    }
                }
            ),
            base_address: 1963261952,
            size: 917504,
            code_file: "C:\\Windows\\System32\\kernel32.dll",
            code_identifier: "590285E9e0000",
            debug_file: "wkernel32.pdb",
            debug_identifier: "D347455996F747D6BF43C176B2171E681"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "287b19c3-9209-4a2b-bb8f-bcc37f411b11",
                        appendix: 1
                    }
                }
            ),
            base_address: 1964179456,
            size: 368640,
            code_file: "C:\\Windows\\System32\\bcryptPrimitives.dll",
            code_identifier: "59B0DF8F5a000",
            debug_file: "bcryptprimitives.pdb",
            debug_identifier: "287B19C392094A2BBB8FBCC37F411B111"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "ae131c67-27a7-4fa1-9916-b5a4aef41190",
                        appendix: 1
                    }
                }
            ),
            base_address: 1971388416,
            size: 790528,
            code_file: "C:\\Windows\\System32\\rpcrt4.dll",
            code_identifier: "5A49BB75c1000",
            debug_file: "wrpcrt4.pdb",
            debug_identifier: "AE131C6727A74FA19916B5A4AEF411901"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "6bedcbce-0a3a-40e9-8040-81c2c8c6cc2f",
                        appendix: 1
                    }
                }
            ),
            base_address: 1972305920,
            size: 917504,
            code_file: "C:\\Windows\\System32\\ucrtbase.dll",
            code_identifier: "59BF2B5Ae0000",
            debug_file: "ucrtbase.pdb",
            debug_identifier: "6BEDCBCE0A3A40E9804081C2C8C6CC2F1"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "8462294a-c645-402d-ac82-a4e95f61ddf9",
                        appendix: 1
                    }
                }
            ),
            base_address: 1994063872,
            size: 1708032,
            code_file: "C:\\Windows\\System32\\KERNELBASE.dll",
            code_identifier: "59BF2BCF1a1000",
            debug_file: "wkernelbase.pdb",
            debug_identifier: "8462294AC645402DAC82A4E95F61DDF91"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "971f98e5-ce60-41ff-b2d7-235bbeb34578",
                        appendix: 1
                    }
                }
            ),
            base_address: 1997996032,
            size: 1585152,
            code_file: "C:\\Windows\\System32\\ntdll.dll",
            code_identifier: "59B0D8F3183000",
            debug_file: "wntdll.pdb",
            debug_identifier: "971F98E5CE6041FFB2D7235BBEB345781"
        }
    ]
}
